codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,85 +315,98 @@ private Http2Stream requireStream(int streamId) {
      * </p>
      */
     private final class FlowControlledData extends FlowControlledBase {
-        private ByteBuf data;
+        private final ArrayDeque<Object> bufsAndPromises = new ArrayDeque<Object>();
```

Comment:
I think there is an expected order that the promise is first, followed by the buffer to which that promise applies.  Maybe add a comment?